source/s3_client.c

```diff
@@ -336,6 +342,33 @@ PyObject *aws_py_s3_client_new(PyObject *self, PyObject *args) {
     s3_client->py_core = py_core;
     Py_INCREF(s3_client->py_core);
 
+    if (network_interface_names_py != Py_None) {
+        if (!PySequence_Check(network_interface_names_py)) {
+            PyErr_SetString(PyExc_TypeError, "Expected network_interface_names to be a sequence.");
+            goto cleanup;
+        }
+        Py_ssize_t list_size = PySequence_Size(network_interface_names_py);
+        if (list_size < 0) {
+            goto cleanup;
+        }
+        num_network_interface_names = (size_t)list_size;
+        network_interface_names =
+            aws_mem_calloc(allocator, num_network_interface_names, sizeof(struct aws_byte_cursor));
+        for (size_t i = 0; i < num_network_interface_names; ++i) {
+            PyObject *str_obj = PySequence_GetItem(network_interface_names_py, i); /* New reference */
+            if (!str_obj) {
+                PyErr_SetString(PyExc_TypeError, "Expected network_interface_names elements to be non-null.");
+                goto cleanup;
+            }
+            network_interface_names[i] = aws_byte_cursor_from_pyunicode(str_obj);
+            Py_DECREF(str_obj);
```

Comment:
😬 now that I'm thinking about it, this isn't 1000% safe either

if (in theory) the sequence wasn't a tuple or list but something with a custom `__getitem__()` that returned a new object, then when we decref it here, it would get cleaned up, and our `char *` would be pointing at discarded memory

I looked at how PySequence_Fast_GET_ITEM() could get away with a borrowed reference, and it does it by ensuring a list or tuple was passed it, and if it was something else, it creates a list instead so it can be sure the items live at least as long as whatever's returned ([see code](https://github.com/python/cpython/blob/376d66eb5080cf8076d767b0916c103463343963/Objects/abstract.c#L2015))

anyway blugh ugh ugh ugh ugh I realize this bug exists elsewhere in our C-bindings

maybe keep these items alive until `cleanup`, similar to the array of cursors. Like:
- declare them before any gotos
    ```
    struct aws_byte_cursor *network_interface_name_cursors = NULL;
    PyObject **network_interface_name_pyobjects = NULL;
    ```
- `aws_mem_calloc()` them at the same time
- `network_interface_name_pyobjects[i] = PySequence_GetItem(network_interface_names_py, i); /* New reference */`
- then in cleanup, loop through and Py_XDECREF() them all before aws_mem_release of the tmp array

Reply:
I'm not TRYING to make your life miserable. I'm just like "oh hey PySequence exists, let's use it" ... "uh oh"

Another option is doing like PySequence_Fast .. and turning it into a `list` (if necessary) before processing it. In python we could just be like:
```
if network_interface_names is not None:
    # ensure this is a list, so it's simpler to process in C
    if not isinstance(network_interface_names, list):
        network_interface_names = list(network_interface_names)
```

Reply:
Thanks, I have updated it to the second option.